scripts/src/chartprreview/chartprreview.py

```diff
@@ -300,7 +302,23 @@ def check_report_success(directory, api_url, report_path, report_info_path, vers
             write_error_log(directory, msg)
             sys.exit(1)
 
+def verify_package_digest(url,report):
+    print("[INFO] check package digest.")
+
+    response = requests.get(url, allow_redirects=True)
+    target_digest = hashlib.sha256(response.content).hexdigest()
 
+    found,report_data = verifier_report.get_report_data(report)
+    if found:
+        pkg_digest = verifier_report.get_package_digest(report_data)
+
+    if target_digest:
+        if pkg_digest and pkg_digest != target_digest:
+            # Digest was passed and computed but differ
+            raise Exception("Found an integrity issue. SHA256 digest passed does not match SHA256 digest computed.")
+    elif not pkg_digest:
+        # Digest was not passed and could not be computed
+        raise Exception("Was unable to compute SHA256 digest, please ensure chart url points to a chart package.")
 
```

Comment:
I'd simplify the flow a little bit, including an additional exception when report data can't be found in the report.

```python
def verify_package_digest(url, report):
    print("[INFO] check package digest.")

    response = requests.get(url, allow_redirects=True)
    target_digest = hashlib.sha256(response.content).hexdigest()

    if not target_digest:
        raise Exception("Was unable to compute SHA256 digest, please ensure chart url points to a chart package.")

    found, report_data = verifier_report.get_report_data(report)
    if not found:
        raise Exception("Was unable to collect report data, can't compute SHA256 digest without it.")

    pkg_digest = verifier_report.get_package_digest(report_data)

    if pkg_digest != target_digest:
        raise Exception("Found an integrity issue. SHA256 digest passed does not match SHA256 digest computed.")
```

Reply:
I've added the changes you've pointed out and an explicit check for when the url can't be retrieved, together with documentation regarding the conditions you've mentioned:

```python
def verify_package_digest(url, report):
    """
    Verifies whether the artifact found at given `url` and the digest contained 
    in the given `report`. It raises an exception when:

    - The given report doesn't contain valid report data.
    - The report data contained in the given report doesn't contain a valid package digest
    - The given url's digest is different than the digest available in report data.
    """
    print("[INFO] check package digest.")

    found, report_data = verifier_report.get_report_data(report)
    
    if not found:
        raise Exception("Was unable to collect report data, can't compute SHA256 digest without it.")

    # pkg_digest can be None
    pkg_digest = verifier_report.get_package_digest(report_data)

    # raises exception if pkg_digest is None or ""
    if not pkg_digest:
        raise Exception("Was unable to collect package digest, can't compute SHA256 digest without it.)

    # download the chart
    response = requests.get(url, allow_redirects=True)

    if response.status_code == 200:
        # calculate the chart's digest from the response contents if the given url is reachable
        target_digest = hashlib.sha256(response.content).hexdigest()

        # when both pkg_digest and target_digest are set, and they're not equal raise an exception.
        if target_digest and pkg_digest != target_digest:
            raise Exception("Found an integrity issue. SHA256 digest passed does not match SHA256 digest computed.")
    else:
        # no-op when the url's SHA256 can't be computed but the report contains a digest.
        pass
```

Reply:
Not trying to introduce any new errors here compared to previous code later in the process which is from where I copied the logic: https://github.com/openshift-helm-charts/development/blob/a8c241f353eb332a02cfbb66842fde96766cf815/scripts/src/chartrepomanager/chartrepomanager.py#L170

Not getting a pkg_digest is only an error if we also cannot get a package digest. With this in mind I think the original logic is fine except for not checking for a 200 response which I now added.